src/System.Windows.Forms/src/System/Windows/Forms/OLE/DataObject.ComposedDataObject.NativeDataObjectToWinFormsAdapter.cs

```
@@ -440,42 +531,71 @@ private static unsafe string ReadUtf8StringFromHGLOBAL(HGLOBAL hglobal)
+                    {
+                        data = t;
+                    }
 
-                object? originalData = data;
+                    return data is object;
+                }
 
+                T? originalData = data;
                 // Try to find a mapped format that works better.
                 foreach (string mappedFormat in mappedFormats)
                 {
                     if (!format.Equals(mappedFormat))
                     {
-                        data = GetObjectFromDataObject(nativeDataObject, mappedFormat, out doNotContinue);
+                        data = GetObjectFromDataObject<T>(nativeDataObject, mappedFormat, resolver, out doNotContinue, out stream);
                         if (doNotContinue)
                         {
                             break;
                         }
 
                         if (data is not null and not MemoryStream)
                         {
-                            return data;
+                            return true;
                         }
                     }
                 }
 
-                return originalData ?? data;
+                data = originalData ?? data;
+                return data is object;
+            }
+
+            private static bool IsUnbounded(Type type) => type.IsInterface || type.IsAbstract || type == typeof(object);
+
+            #region IDataObject
+
+            bool IDataObject.TryGetData<[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] T>(
+                string format,
+                Func<TypeName, Type> resolver,
+                bool autoConvert,
+                [NotNullWhen(true), MaybeNullWhen(false)] out T data)
+            {
+                data = default;
+                if (Clipboard.UnboundedResolver != resolver && IsUnbounded(typeof(T)))
+                {
+                    // TanyaSo - should this be a not supported exception?
+                    return false;
```

My 2 cents here is that we should throw instead of failing silently in this case since it is not very intuitive that the T should not be an unbounded type.